pkg/cli/cli.go

```diff
@@ -0,0 +1,207 @@
+
+package cli
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+
+	apiv1 "github.com/google/cloud-android-orchestration/api/v1"
+
+	"github.com/spf13/cobra"
+)
+
+// Groups streams for standard IO.
+type IOStreams struct {
+	In     io.Reader
+	Out    io.Writer
+	ErrOut io.Writer
+}
+
+type CommandOptions struct {
+	IOStreams
+	Args []string
+}
+
+type CVDRemoteCommand struct {
+	command *cobra.Command
+}
+
+func NewCVDRemoteCommand() *CVDRemoteCommand {
+	opts := &CommandOptions{
+		IOStreams: IOStreams{In: os.Stdin, Out: os.Stdout, ErrOut: os.Stderr},
+		Args:      os.Args[1:],
+	}
+	return NewCVDRemoteCommandWithArgs(opts)
+}
+
+const (
+	serviceURLFlag = "service_url"
+)
+
+type configFlags struct {
+	ServiceURL string
+}
+
+func NewCVDRemoteCommandWithArgs(o *CommandOptions) *CVDRemoteCommand {
+	var configFlags = new(configFlags)
+	hostCmd := &cobra.Command{
+		Use:   "host",
+		Short: "Work with hosts",
+	}
+	hostCmd.AddCommand(newCreateHostCommand())
+	rootCmd := &cobra.Command{
+		Use:               "cvdremote",
+		Short:             "Manages Cuttlefish Virtual Devices (CVDs) in the cloud.",
+		SilenceUsage:      true,
+		SilenceErrors:     true,
+		CompletionOptions: cobra.CompletionOptions{DisableDefaultCmd: true},
+	}
+	rootCmd.SetArgs(o.Args)
+	rootCmd.SetOut(o.IOStreams.Out)
+	rootCmd.SetErr(o.IOStreams.ErrOut)
+	rootCmd.PersistentFlags().StringVar(&configFlags.ServiceURL, serviceURLFlag, "",
+		"Cloud orchestration service url.")
+	rootCmd.MarkPersistentFlagRequired(serviceURLFlag)
+	// Do not show a `help` command, users have always the `-h` and `--help` flags for help purpose.
+	rootCmd.SetHelpCommand(&cobra.Command{Hidden: true})
+	rootCmd.AddCommand(hostCmd)
+	return &CVDRemoteCommand{rootCmd}
+}
+
+// ExecuteNoErrOutput is a version of Execute which returns the command error instead of
+// printing it and do not exit in case of error.
+func (c *CVDRemoteCommand) ExecuteNoErrOutput() error {
+	return c.command.Execute()
+}
+
+func (c *CVDRemoteCommand) Execute() {
+	if err := c.ExecuteNoErrOutput(); err != nil {
+		c.command.PrintErrln(err)
+		os.Exit(1)
+	}
+}
+
+const (
+	machineTypeFlag    = "machine_type"
+	minCpuPlatformFlag = "min_cpu_platform"
+)
+
+type createHostCommandRunner struct {
+	MachineType    *string
+	MinCPUPlatform *string
+	Command        *cobra.Command
+}
+
+func newCreateHostCommand() *cobra.Command {
+	var machineType string
+	var minCPUPlatform string
+	runner := new(createHostCommandRunner)
+	cmd := &cobra.Command{
+		Use:   "create",
+		Short: "Creates a host.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runner.Run()
+		},
+	}
+	cmd.Flags().StringVar(&machineType, machineTypeFlag, "n1-standard-4", "Machine type.")
+	cmd.Flags().StringVar(&minCPUPlatform, minCpuPlatformFlag, "Intel Haswell",
+		"Specifies a minimum CPU platform for the VM instance.")
+	runner.MachineType = &machineType
+	runner.MinCPUPlatform = &minCPUPlatform
+	runner.Command = cmd
```

Comment:
If you pass the `cmd` as a parameter to `runner.Run()` you won't need to hold a reference to it from the runner struct, which avoids the circular dependency between runner and command.

Without that circular dependency you can instantiate the entire runner before the command.

Also, make `MachineType` and `MinCPUPlatform` strings instead of pointers, with the command taking its addresses in the flag specification.

Reply:
good catch! done as you mentioned. removed the handling of `MachineType` and `MinCPUPlatform` as these are not yet handled